asmcomp/cmmgen.ml

```diff
@@ -178,7 +178,10 @@ let rec expr_size env = function
 let transl_constant dbg = function
   | Uconst_int n ->
       int_const dbg n
-  | Uconst_ref (label, _) ->
+  | Uconst_ref (label, def_opt) ->
+      Option.iter
+        (fun def -> Cmmgen_state.add_structured_constant label def)
+        def_opt;
```

Comment:
I may be mistaken, but it looks like you're assuming `transl_constant` is called  on all constants of interest before the corresponding symbols are looked up using `Cmmgen_state.structured_constant_of_sym`, e.g. in `Cmmgen.isunboxed_number_cmm`.  Is that so?

Reply:
Yes. #9310 also adds constant information in `Cmmgen.box_int`, I'm not sure it's needed as unlike the missing constant from other modules, these cases were never unboxed before #2165. But I can patch it too. In all other cases were a `Cconst_symbol` term is created, it doesn't correspond to a known structured constant (thanks to @stedolan's patch, it's easy to check: all `Cconst_symbol` terms appear in the patch, and all the places where his patch adds `None` can be ignored, so that leaves only the `transl_constant` and `box_int` cases).

All calls to `Cmmgen_state.structured_constant_of_sym` work on symbols coming from `Cconst_symbol` terms. So by adding the symbols to the hash table whenever a `Cconst_symbol` term that may correspond to a boxed number is created, we ensure that when we do the lookup the information is there (this is exactly the same reasoning as #9310, by the way, except for the place the information is stored).
In theory, we could even skip the initialisation of the table using `Cmmgen_state.set_structured_constants` (in `Cmmgen.compunit`), as all constants that we care about will also be present as `Uconst_ref` terms. But if for some reason some constants appear as `Uconst_ref(sym, None)` but have information in the `constants` part of `Cmmgen.compunit`'s argument then this could make a difference. #9310 doesn't handle that, but I don't think it can occur in practice, neither with `Closure` nor with `Flambda`.